openssl-dynamic/src/main/c/ssl.c

```diff
@@ -221,6 +234,290 @@ static const jint supported_ssl_opts = 0
+        // The internal SSL buffer exists because a SSL_read operation may actually write data (e.g. handshake).
+        return writeAmount;
+    }
+
+    if (bioUserData->buffer == NULL || bioUserData->bufferLength == 0) {
+        BIO_set_retry_write(bio); /* no buffer to write into */
+        return -1;
+    }
+
+    // First check if we need to drain data queued in the internal SSL buffer.
+    if (bioUserData->nonApplicationBufferLength != 0) {
+        writeAmount = tcn_flush_sslbuffer_to_bytebuffer(bioUserData);
+    }
+
+    // Next write "in" into what ever space the ByteBuffer has available.
+    writeChunk = TCN_MIN(bioUserData->bufferLength, (jint) inl) * sizeof(char);
+
+#ifdef NETTY_TCNATIVE_BIO_DEBUG
+    fprintf(stderr, "tcn_write_to_bytebuffer2 writeChunk %d\n", writeChunk);
+#endif
+
+    memcpy(bioUserData->buffer, in, (size_t) writeChunk);
+    bioUserData->bufferLength -= writeChunk;
+    bioUserData->buffer += writeChunk; // Pointer arithmetic based on char* type
+
+    return writeAmount + writeChunk;
+}
+
+static jint tcn_read_from_bytebuffer(BIO* bio, char *out, int outl) {
+    jint readAmount;
+    struct TCN_bio_bytebuffer* bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio);
+    TCN_ASSERT(bioUserData != NULL);
+
+#ifdef NETTY_TCNATIVE_BIO_DEBUG
+    fprintf(stderr, "tcn_read_from_bytebuffer bioUserData->bufferIsSSLWriteSink %d outl %d [%.*s]\n", bioUserData->bufferIsSSLWriteSink, outl, outl, out);
+#endif
+
+    if (out == NULL || outl <= 0) {
+        return 0;
+    }
+
+    if (bioUserData->bufferIsSSLWriteSink || bioUserData->buffer == NULL || bioUserData->bufferLength == 0) {
+        // During handshake this may happen, and it means we are not setup to read yet.
+        BIO_set_retry_read(bio);
+        return -1;
+    }
+
+    readAmount = TCN_MIN(bioUserData->bufferLength, (jint) outl) * sizeof(char);
+
+#ifdef NETTY_TCNATIVE_BIO_DEBUG
+    fprintf(stderr, "tcn_read_from_bytebuffer readAmount %d\n", readAmount);
+#endif
+
+    memcpy(out, bioUserData->buffer, (size_t) readAmount);
+    bioUserData->bufferLength -= readAmount;
+    bioUserData->buffer += readAmount; // Pointer arithmetic based on char* type
+
+    return readAmount;
+}
+
+static int bio_java_bytebuffer_create(BIO* bio) {
+    struct TCN_bio_bytebuffer* bioUserData = (struct TCN_bio_bytebuffer*) OPENSSL_malloc(sizeof(struct TCN_bio_bytebuffer));
+    if (bioUserData == NULL) {
+        return 0;
+    }
+    // The actual ByteBuffer is set from java and may be swapped out for each operation.
+    bioUserData->buffer = NULL;
+    bioUserData->bufferLength = 0;
+    bioUserData->bufferIsSSLWriteSink = false;
+    bioUserData->nonApplicationBuffer = NULL;
+    bioUserData->nonApplicationBufferSize = 0;
+    bioUserData->nonApplicationBufferOffset = 0;
+    bioUserData->nonApplicationBufferLength = 0;
+
+    BIO_set_data(bio, bioUserData);
+
+    // https://www.openssl.org/docs/man1.1.0/crypto/BIO_set_data.html
```

Comment:
this actually link was meant to help provide motivation for the call to `BIO_set_shutdown` and `BIO_set_init` which is necessary to inform OpenSSL of the BIO's initialization state. I can add comments to this effect but I think the comment belongs here.